### ISSUE TITLE
Boolean created by comparing two not null timestamps is incorrectly nullable.

### ISSUE THREAD

**Issue by kfajdsl** [ts-safeql/safeql#199]
**Describe the bug**
When comparing two not null timestamps, the result will never be null, but SafeQL interprets it at nullable.

**To Reproduce**
Create a table with a `NOT NULL` `timestampz` column:
```sql
CREATE TABLE token (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  expires_at TIMESTAMPZ NOT NULL DEFAULT NOW()
);
```

Query that table, checking if the timestamp is in the past:
```typescript
  const tokenExpired = (await sql<{ expired: boolean; }[]>{`
    SELECT expires_at < NOW() AS expired
    FROM token
    WHERE id = ${tokenId}
  `)[0]?.expired;
```

SafeQL now reports an error, 
```
Query has incorrect type annotation.
	Expected: { expired: boolean; }
	Actual: { expired: boolean | null; }[]
```

**Expected behavior**
SafeQL doesn't report an error.

**Desktop (please complete the following information):**
 - OS: macOS
 - PostgreSQL version: 16.1

**Comment by kfajdsl** [comment]
I also get this problem with a `timestampz` that is created as the result of adding an `interval` to a `timestampz`.

**Comment by Newbie012** [comment]
Good catch. As a workaround for now, you could wrap the condition with `coalesce`. I will push a new version soon that should fix it.

**Comment by karlhorky** [comment]
I can confirm that this happens, also for `IS NOT NULL` expressions with `LEFT JOIN`ed tables:

```ts
await sql<{ id: number, first_name: string, is_mammal: false }[]>`
  SELECT
    animals.id,
    animals.first_name,
    mammals.id IS NOT NULL AS is_mammal
  FROM
    animals
    LEFT JOIN mammals ON mammals.animal_id = animals.id
`;
```

**Comment by karlhorky** [comment]
## Workaround

The [`coalesce` workaround mentioned above](https://github.com/ts-safeql/safeql/issues/199#issuecomment-1891102864) seems to work, although needs also extra `::bool` casting inside the `coalesce()`:

```ts
await sql<{ id: number, first_name: string, is_mammal: boolean }[]>`
  SELECT
    animals.id,
    animals.first_name,
    coalesce(
      (mammals.id IS NOT NULL)::bool,
      FALSE
    ) AS is_mammal
  FROM
    animals
    LEFT JOIN mammals ON mammals.animal_id = animals.id
`;
```

**Comment by karlhorky** [comment]
@Newbie012 would this change be a small special-casing change, similar to this commit here?

- https://github.com/ts-safeql/safeql/commit/07c746a380e43efd84411a0220c8bdedca1a834b

cc @timvandam in case you'd like to try out a small PR

**Comment by karlhorky** [comment]
> I can confirm that this happens, also for `IS NOT NULL` expressions with `LEFT JOIN`ed tables:

Ok, it seems this is a very different case when it comes to the code.

I've tried out my hand at a first PR, not sure if this makes sense like this:

- https://github.com/ts-safeql/safeql/pull/265

**Comment by karlhorky** [comment]
I'll try to open another PR today to fix expressions with comparison operators (`<`, `>`, `=`, more?) being recognized as nullable.

**Comment by Newbie012** [comment]
@karlhorky The above-mentioned example is a bit more complex since you can't tell whether it's nullable or not just by the AST. If the column reference is nullable, then `col > now()` could be null. This missing context needs to be passed down to the nullability checker.
